### PR TITLE
v3: improve `proxy.DomainForward` method signature

### PR DESCRIPTION
**Pull request by ryanbekhen**
## Description

I want to patch proxy.DomainForward which only works on one domain.

Case:
```go
app.Use(proxy.DomainForward("a.example.com", "http://localhost:3000")
app.Use(proxy.DomainForward("b.example.com", "http://localhost:4000")
```
a.example.com will not work and only b.example.com will work.

Problem Solving:
i made it easier as below.

```go
app.Use(DomainForward(map[string]string{
  "a.example.com":  "http://localhost:3000",
  "b.example.com":  "http://localhost:4000",
}))
```

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [ ] For new functionalities I follow the inspiration of the express js framework and built them similar in usage
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation - https://github.com/gofiber/docs for https://docs.gofiber.io/
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] If new dependencies exist, I have checked that they are really necessary and agreed with the maintainers/community (we want to have as few dependencies as possible)
- [x] I tried to make my code as fast as possible with as few allocations as possible
- [ ] For new code I have written benchmarks so that they can be analyzed and improved

## Commit formatting:

Use emojis on commit messages so it provides an easy way of identifying the purpose or intention of a commit. Check out the emoji cheatsheet here: https://gitmoji.carloscuesta.me/

### CODE REVIEW

**Comment by ryanbekhen** [comment]
@ReneWerner87 please review later.

**Comment by ReneWerner87** [comment]
But this is a breaking change
We have to think before we design exposed methods for the library

**Comment by ryanbekhen** [comment]
@ReneWerner87 so what should I do if the PR labeled `Breaking Change`.

**Comment by ReneWerner87** [comment]
We can not accept the improvent in v2
One possibility would be to try to make it backward compatible

**Comment by ryanbekhen** [comment]
@ReneWerner87  please look at commit  14cf4b67366b4092de80f688c5dd4da7a08dbacf, I just changed the way to use DomainForward so I can add domains. Does this include Breaking change?

**Comment by ReneWerner87** [comment]
The first parameter of DomainForward is a complete different datatype
This is a breaking change, without compatibility to the usage before

**Comment by ryanbekhen** [comment]
@ReneWerner87  Ok thanks for the info. Does this mean it will be merged in the next minor version?

**Comment by ReneWerner87** [comment]
This means that we will be able to adjust it in the next major version at the earliest.

This will probably come out in a few months

**Comment by ryanbekhen** [comment]
@ReneWerner87 maybe I can just add documentation when using multiple domains.

**Comment by ReneWerner87** [comment]
Ok, documentation changes are okay

**Comment by ryanbekhen** [comment]
ok I will add the documentation later

**Comment by efectn** [comment]
@ryanbekhen can you create/edit a PR that targets v3 branch. I think this is nice change for v3.

**Comment by ryanbekhen** [comment]
@efectn Yes, I can.  So, when can I add this feature?

**Comment by efectn** [comment]
> @efectn Yes, I can. So, when can I add this feature?

When you want. I merged master into the v3 beta branch

**Comment by ryanbekhen** [comment]
@efectn Previously I added 2 features, namely proxy.DomainForward and proxy.BalancerForward. Do I add these two features to v3 or only proxy.DomainForward?

**Comment by efectn** [comment]
> @efectn Previously I added 2 features, namely proxy.DomainForward and proxy.BalancerForward. Do I add these two features to v3 or only proxy.DomainForward?

They are already exists on v3-beta -> https://github.com/gofiber/fiber/blob/v3-beta/middleware/proxy/proxy.go#L176

**Comment by ryanbekhen** [comment]
@efectn I have changed the target branch to v3-beta

File: middleware/proxy/proxy.go [hunk 1]
```diff
@@ -173,11 +173,15 @@ func getScheme(uri []byte) []byte {
 
 // DomainForward performs an http request based on the given domain and populates the given http response.
 // This method will return an fiber.Handler
-func DomainForward(hostname, addr string, clients ...*fasthttp.Client) fiber.Handler {
-	return func(c fiber.Ctx) error {
-		host := string(c.Request().Host())
-		if host == hostname {
-			return Do(c, addr+c.OriginalURL(), clients...)
+func DomainForward(config DomainForwardConfig, clients ...*fasthttp.Client) fiber.Handler {
+	return func(c *fiber.Ctx) error {
```
Comment (efectn) [inline middleware/proxy/proxy.go:177]: Remove the pointer here

File: middleware/proxy/proxy_test.go [hunk 2]
```diff
@@ -497,51 +497,54 @@ func Test_ProxyBalancer_Custom_Client(t *testing.T) {
...
 	t.Parallel()
-	ln, err := net.Listen(fiber.NetworkTCP4, "127.0.0.1:0")
-	require.NoError(t, err)
-	app := fiber.New()
-
-	// target server
-	ln1, err := net.Listen(fiber.NetworkTCP4, "127.0.0.1:0")
-	require.NoError(t, err)
-	app1 := fiber.New()
+	_, addr1 := createProxyTestServer(t, func(c *fiber.Ctx) error {
```
Comment (efectn) [inline middleware/proxy/proxy_test.go:503]: Here too

File: middleware/proxy/proxy_test.go [hunk 3]
```diff
@@ -497,51 +497,54 @@ func Test_ProxyBalancer_Custom_Client(t *testing.T) {
...
+		return c.SendString("a.example.com")
+	})
 
-	app1.Get("/test", func(c fiber.Ctx) error {
-		return c.SendString("test_local_client:" + c.Query("query_test"))
+	_, addr2 := createProxyTestServer(t, func(c *fiber.Ctx) error {
```
Comment (efectn) [inline middleware/proxy/proxy_test.go:507]: Here

File: middleware/proxy/proxy_test.go [hunk 4]
```diff
@@ -497,51 +497,54 @@ func Test_ProxyBalancer_Custom_Client(t *testing.T) {
...
+	utils.AssertEqual(t, string(b1), "a.example.com")
+
+	req2 := httptest.NewRequest(fiber.MethodGet, "/", nil)
+	req2.Host = "b.example.com"
+	resp2, err2 := app.Test(req2)
+	utils.AssertEqual(t, nil, err2)
+	utils.AssertEqual(t, fiber.StatusOK, resp2.StatusCode)
+
+	b2, err2 := io.ReadAll(resp2.Body)
+	utils.AssertEqual(t, nil, err2)
```
Comment (efectn) [inline middleware/proxy/proxy_test.go:535]: Change assertequals with require

**Review by efectn — CHANGES_REQUESTED** [review]
There are some linting errors that unrelated to this PR. I'll fix them later.

**Comment by efectn** [comment]
@ryanbekhen can you check my last comments

**Comment by ryanbekhen** [comment]
@efectn i have changed assertequals to require in last commit

**Review by efectn — APPROVED** [review]

**Comment by ryanbekhen** [comment]
@efectn I'm going to update the readme on the proxy as the signature section hasn't been updated yet.

**Review by efectn — APPROVED** [review]

**Pull request merged**